packages/core/src/storage/vectorStore/TiDBVectorStore.ts

```diff
@@ -0,0 +1,322 @@
+    return db.query(sql, [this.namespace]);
+  }
+
+  private getDataToInsert(embeddingResults: BaseNode<Metadata>[]) {
+    const result = [];
+    for (let index = 0; index < embeddingResults.length; index++) {
+      const row = embeddingResults[index];
+
+      const id: any = row.id_.length ? row.id_ : null;
+      const externalId = "";
+      const document = row.getContent(MetadataMode.EMBED);
+      const meta = row.metadata || {};
+      meta.create_date = new Date();
+      const embeddings = this.vectorToSQL(row.getEmbedding());
+
+      const params = [
+        this.namespace,
+        id,
+        externalId,
+        document,
+        meta,
+        embeddings,
+      ];
+
+      result.push(params);
+    }
+    return result;
+  }
+
+  /**
+   * Adds vector record(s) to the table.
+   * @param embeddingResults The Nodes to be inserted, optionally including metadata tuples.
+   * @returns A list of zero or more id values for the created records.
+   */
+  async add(embeddingResults: BaseNode<Metadata>[]): Promise<string[]> {
+    if (embeddingResults.length == 0) {
+      console.debug("Empty list sent to TiDBVectorStore::add");
+      return Promise.resolve([]);
+    }
+
+    const sql: string = `INSERT INTO ${this.tableName} (namespace, id, external_id, document, metadata, embeddings) VALUES ($1, $2, $3, $4, $5, $6)`;
+    const db = await this.getDb();
+    const data = this.getDataToInsert(embeddingResults);
+
+    const ret: string[] = [];
+    for (let index = 0; index < data.length; index++) {
+      const params = data[index];
+      try {
+        const [rows] = await db.query<DocumentEmbedding[]>(sql, params);
+        if (rows.length) {
+          const id = rows[0].id as string;
+          ret.push(id);
+        }
+      } catch (err) {
+        const msg = `${err}`;
+        console.log(msg, err);
+      }
+    }
+
+    return Promise.resolve(ret);
+  }
+
+  /**
+   * Deletes a single record from the database by id.
+   * @param refDocId Unique identifier for the record to delete.
+   * @param deleteKwargs Required by VectorStore interface.  Currently ignored.
+   * @returns Promise that resolves if the delete query did not throw an error.
+   */
+  async delete(refDocId: string, deleteKwargs?: any): Promise<void> {
+    const namespaceCriteria = this.namespace.length ? "AND namespace = $2" : "";
+    const sql: string = `DELETE FROM ${this.tableName} WHERE id = $1 ${namespaceCriteria}`;
+    const db = await this.getDb();
+    const params = this.namespace.length
+      ? [refDocId, this.namespace]
+      : [refDocId];
+    await db.query(sql, params);
+    return Promise.resolve();
+  }
+
+  /**
+   * Query the vector store for the closest matching data to the query embeddings
+   * @param query The VectorStoreQuery to be used
+   * @param options Required by VectorStore interface.  Currently ignored.
+   * @returns Zero or more Document instances with data from the vector store.
+   */
+  async query(
```

Comment:
We're using metadataDictToNode and nodeToMetadata from "./utils.js" to de/serialize the node in the metadata. You can reference the other vector store implementations to have a look into this.